Cross-platform media controller for video playback

rh101 opened the pull request:
## Describe your changes

This PR is to add a media controller to video playback, which works across all supported platforms and media engine backends.

- [x] AndroidMediaEngine
- [x] AvfMediaEngine
- [x] MfMediaEngine
- [x] VlcMediaEngine
- [x] WmfMediaEngine 

@halx99 Is the VLC engine only for Linux? Attempting to compile it on Windows ended up with this error:
`3>Error copying directory from "G:/game_engines/axmol/3rdparty/vlc/win/lib/vlc/plugins" to "G:/dev/public/tests/videotest/build_x64/bin/videotest/Debug/plugins".`

I don't see the `vlc` folder in the `axmol/3rdparty` either, and I'm not sure what else I need to do besides enable it via `set(AX_ENABLE_VLC_MEDIA ON CACHE BOOL "" FORCE)`.

By default the player controls are off.  Would it be preferrable to have them disabled by default?

At the moment, to enable or disable them, just use `VideoPlayer::setUserInputEnabled()`. For example:
```
auto* videoPlayer = ui::VideoPlayer::create();
videoPlayer->play();
videoPlayer->setUserInputEnabled(true);
```

The default media controller can also be overriden with a custom version, and applied to the video player via `MediaPlayer::setMediaController(MediaController* controller)`.

The Android implementation required a work-around, since there was no way to retrieve the content duration and current time directly.  It has to happen in callbacks, but I'm not sure if the solution implemented in this PR is the most appropriate, so please review that, specifically the code in `AxmolMediaEngine.java`, method `processVideoFrame`, where it calls back with the current video playback time.

## Issue ticket number and link
#1811 

## Checklist before requesting a review
### For each PR
- [ ] Add Copyright if it missed:   
      - `"Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md)."`
- [x] I have performed a self-review of my code.
       
   Optional:
   - [x] I have checked readme and add important infos to this PR.
   - [ ] I have added/adapted some tests too.
          
### For core/new feature PR
- [x] I have checked readme and add important infos to this PR.
- [ ] I have added thorough tests.

rh101 committed:
Initial work on media controller

rh101 committed:
Media player controller for WMF media engine

rh101 committed:
Add getDuration and getCurrentTime methods to media engines

rh101 committed:
Fix compilation

rh101 committed:
Fix Android media timeline issues

rh101 committed:
Fix compilation issue

rh101 committed:
Fix WinRT compilation issue

rh101 committed:
Fix compilation issue

rh101 committed:
Add missing header include

rh101 committed:
Make it simpler to create a custom media controller

rh101 committed:
Ensure the basic media controller widget size is the same size as the media player on initialization

aismann commented:
@rh101 
Please add a copyright header/comment on each file too

rh101 committed:
Add dark media control underlay to give more contrast to the controls
ADd copyright notices

halx99 commented:
> @halx99 Is the VLC engine only for Linux? Attempting to compile it on Windows ended up with this error:
> `3>Error copying directory from "G:/game_engines/axmol/3rdparty/vlc/win/lib/vlc/plugins" to "G:/dev/public/tests/videotest/build_x64/bin/videotest/Debug/plugin`

Yes, it's only for linux, and axmol prebuilt libvlc not work for current axmol version

rh101 committed:
Removing existing media controller from parent if required

rh101 commented:
Currently in the process of testing the Linux version, and fixing a few issues related to content scaling.  I'm trying to convert this PR to a draft with the button on github, but it's not letting me, so for the time being, please don't merge anything in.

rh101 committed:
Adjust for view scaling since embedded resources are created for a scale of 1.

rh101 committed:
Scale central media controls if required on content size change

rh101 committed:
Remove pointless fast forward and fast reverse buttons
Ensure media controller correctly resizes on full-screen media playback

rh101 commented:
Linux version runs, but I'm having issues with missing media player control font characters. The Android version is also missing the play button font character, yet all the other buttons appear correctly, so another issue with the fonts on that system. 

I'm not sure how to resolve this without using image files, but with image files, to not look pixelated, they end up as 3-4K file each, so the resulting base64 encoded PNGs are pretty large.

Any suggestions?

halx99 commented:
> Linux version runs, but I'm having issues with missing media player control font characters. The Android version is also missing the play button font character, yet all the other buttons appear correctly, so another issue with the fonts on that system.
> 
> I'm not sure how to resolve this without using image files, but with image files, to not look pixelated, they end up as 3-4K file each, so the resulting base64 encoded PNGs are pretty large.
> 
> Any suggestions?

Use drawNode to draw control buttons shape?

halx99 committed:
Allow enable vlc on windows

halx99 commented:
> By default the player controls are on. Would it be preferrable to have them disabled by default?

disable by default maybe better

halx99 committed:
Improve media macro detect

halx99 commented:
Maybe the `UIMediaPlayer::setUserInputEnabled` can be removed, instead use new API: setMediaControllerEnabled ?

rh101 committed:
Merge branch 'axmolengine:dev' into generic-media-controller

halx99 commented:
> > Linux version runs, but I'm having issues with missing media player control font characters. The Android version is also missing the play button font character, yet all the other buttons appear correctly, so another issue with the fonts on that system.
> > I'm not sure how to resolve this without using image files, but with image files, to not look pixelated, they end up as 3-4K file each, so the resulting base64 encoded PNGs are pretty large.
> > Any suggestions?
> 
> Use drawNode to draw control buttons shape?

Or embed textures due to the buttons shape is simply. for example progress bar can use one 2x2 white texture, and embed arrow, stop textures

rh101 committed:
Add media controls drawn by DrawNode into single texture
Add fullscreen enter and exit controls
setUserInputEnabled and isUserInputEnabled replaced by setMediaControllerEnabled and isMediaControllerEnabled

rh101 committed:
Update function calls to new names

rh101 committed:
Revert back to setUserInputEnabled and isUserInputEnabled

rh101 committed:
Merge branch 'axmolengine:dev' into generic-media-controller

halx99 commented on core/ui/UIMediaPlayer.cpp:
```diff
@@ -136,20 +138,659 @@ struct PrivateVideoDescriptor
+std::unique_ptr<MediaEngineFactory> _meFactory = MediaEngineFactory::create();
+
+const char* CIRCLE_IMAGE =
+    "iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/"
```
convert to binary maybe better

rh101 replied:
Oh! I actually missed that. We don't need it anymore, since I can just use the DrawNode code to generate it too.

rh101 committed:
Remove embedded circle image and create it with DrawNode

rh101 committed:
Change radius of drawn circle to half the total image size

halx99 commented:
Hi @rh101
@asnagni said can't see controller buttons on iOS, could you take a look? 
you can login to discord to view detail conversations

rh101 commented:
> Hi @rh101 @asnagni said can't see controller buttons on iOS, could you take a look? you can login to discord to view detail conversations

I'll load it up right now and check
